EDU 266: How to schedule a Workflow in Go (#2005)
temporalio/documentation: docs-src/go/how-to-schedule-a-workflow-in-go.md

To backfill a Schedule in Go, use `Backfill()` on the `scheduleHandle`.

## Delete

Deleting a Schedule erases a Schedule from a Workflow.
Deletion does not affect any Workflows started by the Schedule.

To delete a Schedule in Go, use the `Delete()` command on the `scheduleHandle`.

<!--SNIPSTART samples-go-schedule {"selectedLines": ["43-46"]}-->
<!--SNIPEND-->

## Describe

Information about the current Schedule configuration can be retrieved with the `describe` action.
This is helpful when you want to get a detailed view of the Schedule and its associated [Workflow Runs](/concepts/what-is-a-run-id).

To describe a Schedule in Go, use `Describe()` on the `scheduleHandle`.

<!--SNIPSTART samples-go-schedule {"selectedLines": ["107-110"]}-->
<!--SNIPEND-->

## List

When used, `list` returns all available Schedules and their Schedule IDs.

To retrieve a list of Schedules, use `ScheduleClient().List()`.

## Pause/unpause

The `pause` and `unpause` actions enable the start or stop of all future Workflow Runs on a given Schedule.
Pausing a Schedule halts all future Workflow Runs, while unpausing a Schedule allows the Workflow to execute as planned.

> Pausing can be enabled upon creating a Schedule by setting `State.Paused` to `true`.

Reviewer: ```suggestion
Pausing can be enabled when you create a Schedule by setting `State.Paused` to `true`.
```

Alternatively: "When you create a Schedule, you can enable pausing by setting `State.Paused` to `true`."

<!--SNIPSTART samples-go-schedule {"selectedLines": ["85"]}-->
<!--SNIPEND-->

> Pausing can also be done by using `Pause()` on the `scheduleHandle`.

Reviewer: ```suggestion
Pausing can also be done by using `Pause()` on `scheduleHandle`.
```

```go
scheduleHandle.Pause()
```

> To unpause a Schedule, use `Unpause()` on the `scheduleHandle`.

Reviewer: ```suggestion
To unpause a Schedule, use `Unpause()` on `scheduleHandle`.
```

<!--SNIPSTART samples-go-schedule {"selectedLines": ["99-102"]}-->
<!--SNIPEND-->

## Trigger

Triggering a Schedule immediately executes an Action defined in the Schedule.
By default, `trigger` is subject to the Overlap Policy.

> To trigger a Scheduled Workflow Execution in Go, use `Trigger()` on the `scheduleHandle`.

Reviewer: ```suggestion
To trigger a Scheduled Workflow Execution in Go, use `Trigger()` on `scheduleHandle`.
```

---
id: how-to-schedule-a-workflow-in-go
title: How to Schedule a Workflow in Go 
sidebar_label: Schedule
description: Set a Schedule in Go for scheduling Workflows to run at specific times.
tags:
  - go
  - developer-guide
  - how-to
---

Scheduling [Workflows](/concepts/what-is-a-workflow) is crucial to automation.
By scheduling [Workflow Executions](/concepts/what-is-a-workflow-execution), you can reduce manual intervention and ensure timely execution of business processes.

> Temporal allows you to execute multiple Schedule operations, either through code or with the [CLI tool](/concepts/what-is-the-temporal-cli).

Reviewer 2: A user can also use the web UI, might want to call that out.

Read on to find out how to create and edit Schedules.

:::note
Be sure to [enable Schedules in your environment](/concepts/what-is-a-schedule#limitations) before creating any Schedules.
:::

## Create

Schedules are created with the `create` action.

> For each new Schedule, tbe Temporal Server generates a unique Schedule ID.

Reviewer 2: This paragraph says the server generates the schedule ID, that paragraph below says the user must pass an ID. These don't seem consistent

Author: Changed per @cretz correction

To create a Schedule in Go, use `ScheduleClient().Create()` on the [Client](/concepts/what-is-the-temporal-client).
Schedules must be initialized with a Schedule ID, [Spec](/concepts/what-is-a-schedule#spec), and [Action](/concepts/what-is-a-schedule#action) to perform.
Enter these values in `client.Schedule.Options{}`.

<!--SNIPSTART samples-go-schedule-create-delete {"selectedLines": ["26-38"]}-->

[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)

```go
// This schedule ID can be user business logic identifier as well.
scheduleID := "schedule_" + uuid.New()
workflowID := "schedule_workflow_" + uuid.New()

> // Create the schedule, start with no spec so the schedule will not run.

Author: samples-go: https://github.com/temporalio/samples-go/blob/main/schedule/starter/main.go

scheduleHandle, err := c.ScheduleClient().Create(ctx, client.ScheduleOptions{
	ID:   scheduleID,
	Spec: client.ScheduleSpec{},
	Action: &client.ScheduleWorkflowAction{
		ID:        workflowID,
		Workflow:  schedule.SampleScheduleWorkflow,
		TaskQueue: "schedule",
	},
})
if err != nil {
	log.Fatalln("Unable to create schedule", err)
}
```

<!--SNIPEND-->

## Backfill

Backfilling a Schedule executes [Workflow Tasks](/concepts/what-is-a-workflow-task) ahead of the Schedule's specified time range.
This is useful when you need to execute a missed or delayed Action, or if you want to test the Workflow ahead of time.

> To backfill a Schedule in Go, use `Backfill()` on `scheduleHandle`.

Reviewer 2: nit: `scheduleHandle ` -> `ScheduleHandle ` because `ScheduleHandle` is the actual type

## Delete

Deleting a Schedule erases a Schedule from a Workflow.
Deletion does not affect any Workflows started by the Schedule.

To delete a Schedule in Go, use the `Delete()` command on `scheduleHandle`.

<!--SNIPSTART samples-go-schedule-create-delete {"selectedLines": ["42-49"]}-->

[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)

```go

> // Delete the schedule once the sample is done

Author: Can snip around that

> // Delete the schedule once the sample is done

Author: Again, got it from samples-go: https://github.com/temporalio/samples-go/blob/main/schedule/starter/main.go

defer func() {
	log.Println("Deleting schedule", "ScheduleID", scheduleHandle.GetID())

> 	err = scheduleHandle.Delete(ctx)

Reviewer 3: With these samples showing how to use a `scheduleHandle`, we should show them how to obtain it (using the one from create is often not a realistic use even though we do it in the sample).

Reviewer 3: > How do we obtain it?

Via `ScheduleClient.GetHandle`

> I am working from what was done in code

You are working with what was done in a single sample that creates and deletes in the same bit of code. But our docs shouldn't expect people would use it that way. Sample code and docs code can often have mismatched purposes. In this case, samples show you how to do everything together at once and docs separate them into standalone sections. These are incompatible in this case.

> If there's another place mentioning the use of scheduleHandle, please let me know

You may have to write it.

Author: Very well. I'm going to go ahead and put in code blocks in lieu of the code snippets. Once I migrate this over to dacx format, then I'll write code for a use case.

Reviewer 3: May want to confirm with @flossypurse. Maybe updating the sample to break out the steps makes more sense. That's what was done in the Python schedule sample.

To retrieve a list of Schedules, use `ScheduleClient().List()`.

## Pause/unpause

The `pause` and `unpause` actions enable starting or stopping all future Workflow Runs on a given Schedule.
Pausing a Schedule halts all future Workflow Runs; unpausing a Schedule allows the Workflow to execute as planned.

Pausing can be enabled when you create a Schedule by setting `State.Paused` to `true`.

<!--SNIPSTART samples-go-schedule-update {"selectedLines": ["90-91"]}-->

[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)

```go
// Start the schedule paused to demonstrate how to unpause a schedule
schedule.Description.Schedule.State.Paused = true
```

<!--SNIPEND-->

Pausing can also be done by using `Pause()` on `scheduleHandle`.

```go
scheduleHandle.Pause()
```

To unpause a Schedule, use `Unpause()` on `scheduleHandle`.

<!--SNIPSTART samples-go-schedule-unpause-describe {"selectedLines": ["107-110"]}-->

[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)

```go
// Unpause schedule
log.Println("Unpausing schedule", "ScheduleID", scheduleHandle.GetID())

> err = scheduleHandle.Unpause(ctx, client.ScheduleUnpauseOptions{})

Reviewer 3: Inconsistent that unpausing has logs and error checks and everything but pausing is just a single call w/ no error check

if err != nil {
	log.Fatalln("Unable to unpause schedule", err)
}
```

<!--SNIPEND-->

## Trigger

Triggering a Schedule immediately executes an Action defined in the Schedule.
By default, `trigger` is subject to the Overlap Policy.

To trigger a Scheduled Workflow Execution in Go, use `Trigger()` on `scheduleHandle`.
Set desired changes in `ScheduleTriggerOptions`.

<!--SNIPSTART samples-go-schedule-trigger {"selectedLines": ["55-57"]}-->

[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)

```go

> 	// Manually trigger the schedule once

Reviewer 3: Inconsistent indentation on snippets (some of these are left aligned, some indented)

	})
	if err != nil {
		log.Fatalln("Unable to trigger schedule", err)
	}
```

<!--SNIPEND-->

## Update

Updating a Schedule changes the configuration of existing Schedules.

Use the `Update()` command to modify an existing Schedule.

<!--SNIPSTART samples-go-schedule-update {"selectedLines": ["65-99"]}-->

[schedule/starter/main.go](https://github.com/temporalio/samples-go/blob/master/schedule/starter/main.go)

```go
	// Update the schedule with a spec so it will run periodically,
	log.Println("Updating schedule", "ScheduleID", scheduleHandle.GetID())
	err = scheduleHandle.Update(ctx, client.ScheduleUpdateOptions{
		DoUpdate: func(schedule client.ScheduleUpdateInput) (*client.ScheduleUpdate, error) {
			schedule.Description.Schedule.Spec = &client.ScheduleSpec{
				// Run the schedule at 5pm on Friday
				Calendars: []client.ScheduleCalendarSpec{
					{
						Hour: []client.ScheduleRange{
							{
								Start: 17,
							},
						},
						DayOfWeek: []client.ScheduleRange{
							{
								Start: 5,
							},
						},
					},
				},
				// Run the schedule every 5s
				Intervals: []client.ScheduleIntervalSpec{
					{
						Every: 5 * time.Second,
					},
				},
			}

> 			// Start the schedule paused to demonstrate how to unpause a schedule

Reviewer 3: This is not starting (may have to go back and fix that comment in samples) and this isn't for demonstrating unpause

Author: https://github.com/temporalio/samples-go/blob/main/schedule/starter/main.go

Author: There's nothing else demonstrating. I'm working with what's provided, so this looks like a good time to go over the current code samples.

Reviewer 3: :+1: may have to go back and fix that comment in samples

> There's nothing else demonstrating

Exactly, so should fix that comment

Reviewer 3: Oh I didn't take it that way, you're good! Ideally we can get the code right from the outset, but if not, ok.

			}, nil
		},
	})
	if err != nil {
		log.Fatalln("Unable to update schedule", err)
	}
	// Upause schedule
	log.Println("Unpausing schedule", "ScheduleID", scheduleHandle.GetID())
	err = scheduleHandle.Unpause(ctx, client.ScheduleUnpauseOptions{})
	if err != nil {
		log.Fatalln("Unable to unpause schedule", err)
	}
	// Wait for the schedule to run 10 actions
	log.Println("Waiting for schedule to complete 10 actions", "ScheduleID", scheduleHandle.GetID())

	for {
		description, err := scheduleHandle.Describe(ctx)
		if err != nil {
			log.Fatalln("Unable to describe schedule", err)
		}
		if description.Schedule.State.RemainingActions != 0 {
			log.Println("Schedule has remaining actions", "ScheduleID", scheduleHandle.GetID(), "RemainingActions", description.Schedule.State.RemainingActions)
			time.Sleep(5 * time.Second)
		} else {
			break
		}
	}
}
```

<!--SNIPEND-->

## Describe

You can retrieve information about the current Schedule configuration by using the `describe` action.
This is helpful when you want to get a detailed view of the Schedule and its associated [Workflow Runs](/concepts/what-is-a-run-id).

To describe a Schedule in Go, use `Describe()` on `scheduleHandle`.

> <!--SNIPSTART samples-go-schedule {"selectedLines": ["107-110"]}-->

Author: Looking at it
